### ISSUE TITLE
Bad UI performance on Linux Wayland session

### ISSUE THREAD

**Issue by LDVSOFT** [telegramdesktop/tdesktop#26343]
### Steps to reproduce

I've tried using Telegram Desktop on my Linux desktop running Ubuntu 22.04 with AMD videocard and 4K display, and when running under Wayland the UI is very slow (low fps? high latency?). 

### Expected behaviour

Should run as always.

### Actual behaviour

Way slower than on X11.

### Operating system

Ubuntu 22.04

### Version of Telegram Desktop

4.8.1

### Installation source

Static binary from official website

### Logs

```shell
[2023.05.29 10:47:43] Launched version: 4008001, install beta: [FALSE], alpha: 0, debug mode: [FALSE]
[2023.05.29 10:47:43] Executable dir: /opt/Telegram/, name: Telegram
[2023.05.29 10:47:43] Initial working dir: /home/ldvsoft/.local/share/TelegramDesktop/
[2023.05.29 10:47:43] Working dir: /home/ldvsoft/.local/share/TelegramDesktop/
[2023.05.29 10:47:43] Command line: /opt/Telegram/Telegram -workdir /home/ldvsoft/.local/share/TelegramDesktop/ -autostart
[2023.05.29 10:47:43] Executable path before check: /opt/Telegram/Telegram
[2023.05.29 10:47:43] Logs started
[2023.05.29 10:47:43] Launcher filename: org.telegram.desktop._6feb56bf3e01ae72605dedc0d4727bb7.desktop
[2023.05.29 10:47:49] Connecting local socket to 6feb56bf3e01ae72605dedc0d4727bb7-{87A94AB0-E370-4cde-98D3-ACC110C5967D}...
[2023.05.29 10:47:49] Socket connect error 0, starting server and app...
[2023.05.29 10:47:49] Moved logging from '/home/ldvsoft/.local/share/TelegramDesktop/log_start0.txt' to '/home/ldvsoft/.local/share/TelegramDesktop/log.txt'!
[2023.05.29 10:47:49] Opened '/home/ldvsoft/.local/share/TelegramDesktop/tdata/working' for reading, the previous Telegram Desktop launch was not finished properly :( Crash log size: 0
[2023.05.29 10:47:49] Global devicePixelRatio: 2
[2023.05.29 10:47:49] Primary screen DPI: 96, Base: 96.
[2023.05.29 10:47:49] Computed screen scale: 100
[2023.05.29 10:47:49] DevicePixelRatio: 2
[2023.05.29 10:47:49] ScreenScale: 100
[2023.05.29 10:47:49] System tray available: [TRUE]
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansRegular.ttf' loaded 'DAOpenSansRegular'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirRegular.ttf' loaded 'DAVazirRegular'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansRegularItalic.ttf' loaded 'DAOpenSansRegularItalic'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirRegular.ttf' loaded 'DAVazirRegular'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansSemiboldAsBold.ttf' loaded 'DAOpenSansSemibold'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirMediumAsBold.ttf' loaded 'DAVazirMedium'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansSemiboldItalicAsBold.ttf' loaded 'DAOpenSansSemiboldItalic'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirMediumAsBold.ttf' loaded 'DAVazirMedium'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansSemiboldAsBold.ttf' loaded 'DAOpenSansSemibold'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirMediumAsBold.ttf' loaded 'DAVazirMedium'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAOpenSansSemiboldItalicAsBold.ttf' loaded 'DAOpenSansSemiboldItalic'
[2023.05.29 10:47:49] Font: from ':/gui/fonts/DAVazirMediumAsBold.ttf' loaded 'DAVazirMedium'
[2023.05.29 10:47:49] Icon theme: Yaru
[2023.05.29 10:47:49] Fallback icon theme: 
[2023.05.29 10:47:49] App Info: reading settings...
[2023.05.29 10:47:49] App Info: reading encrypted settings...
[2023.05.29 10:47:50] Lang Info: Loaded cached, keys: 4994
[2023.05.29 10:47:51] OpenAL Logging Level: (not set)
[2023.05.29 10:47:51] Audio Playback Devices: Встроенное аудио Цифровой стерео (IEC958);Navi 21 HDMI Audio [Radeon RX 6800/6800 XT / 6900 XT] Digital Stereo (HDMI 2)
[2023.05.29 10:47:51] Audio Playback Default Device: Встроенное аудио Цифровой стерео (IEC958)
[2023.05.29 10:47:51] Audio Capture Devices: Monitor of Встроенное аудио Цифровой стерео (IEC958);Monitor of Navi 21 HDMI Audio [Radeon RX 6800/6800 XT / 6900 XT] Digital Stereo (HDMI 2)
[2023.05.29 10:47:51] Audio Capture Default Device: Monitor of Встроенное аудио Цифровой стерео (IEC958)
[2023.05.29 10:47:52] OpenGL Profile: Compatibility.
[2023.05.29 10:47:52] OpenGL Renderer: AMD Radeon RX 6650 XT (navi23, LLVM 15.0.7, DRM 3.47, 5.19.0-42-generic)
[2023.05.29 10:47:52] OpenGL Vendor: AMD
[2023.05.29 10:47:52] OpenGL Version: 4.6 (Compatibility Profile) Mesa 22.2.5
[2023.05.29 10:47:52] OpenGL Extensions: GL_ARB_compatibility, GL_EXT_EGL_image_storage, GL_ARB_transform_feedback_overflow_query, GL_EXT_framebuffer_multisample_blit_scaled, GL_ARB_fragment_layer_viewport, GL_EXT_texture_buffer_object, GL_ARB_shader_texture_image_samples, GL_ARB_polygon_offset_clamp, GL_EXT_shadow_funcs, GL_ARB_ES3_1_compatibility, GL_EXT_texture_shared_exponent, GL_ARB_shader_precision, GL_ARB_sample_shading, GL_ARB_shading_language_420pack, GL_ARB_conservative_depth, GL_EXT_subtexture, GL_EXT_framebuffer_sRGB, GL_EXT_stencil_two_side, GL_ARB_timer_query, GL_ARB_uniform_buffer_object, GL_EXT_texture3D, GL_NV_fog_distance, GL_OES_EGL_image, GL_KHR_debug, GL_ARB_get_program_binary, GL_NV_vdpau_interop, GL_EXT_polygon_offset_clamp, GL_ARB_draw_buffers_blend, GL_ARB_texture_rectangle, GL_EXT_framebuffer_blit, GL_ARB_texture_rg, GL_ATI_texture_float, GL_AMD_seamless_cubemap_per_texture, GL_ARB_texture_multisample, GL_AMD_shader_trinary_minmax, GL_ATI_fragment_shader, GL_EXT_transform_feedback, GL_NV_packed_depth_stencil, GL_MESA_texture_signed_rgba, GL_INTEL_blackhole_render, GL_EXT_packed_depth_stencil, GL_ARB_shading_language_include, GL_ATI_draw_buffers, GL_EXT_memory_object_fd, GL_ATI_texture_env_combine3, GL_ARB_shader_viewport_layer_array, GL_EXT_texture_integer, GL_ARB_point_sprite, GL_ARB_framebuffer_object, GL_EXT_shader_image_load_formatted, GL_ARB_texture_stencil8, GL_ARB_texture_env_dot3, GL_ARB_sync, GL_ARB_cull_distance, GL_ARB_parallel_shader_compile, GL_ATI_meminfo, GL_ARB_half_float_pixel, GL_ARB_query_buffer_object, GL_ARB_depth_clamp, GL_ARB_texture_view, GL_ARB_internalformat_query2, GL_ARB_vertex_type_10f_11f_11f_rev, GL_NV_copy_image, GL_AMD_conservative_depth, GL_EXT_point_parameters, GL_EXT_copy_texture, GL_ARB_conditional_render_inverted, GL_EXT_gpu_program_parameters, GL_ARB_gpu_shader5, GL_EXT_EGL_sync, GL_KHR_parallel_shader_compile, GL_EXT_shader_integer_mix, GL_ARB_point_parameters, GL_ARB_texture_query_lod, GL_ARB_ES3_2_compatibility, GL_EXT_framebuffer_multisample, GL_KHR_blend_equation_advanced, GL_ARB_texture_gather, GL_SGIS_texture_border_clamp, GL_EXT_texture_snorm, GL_ARB_copy_image, GL_ARB_shader_group_vote, GL_KHR_robust_buffer_access_behavior, GL_EXT_shader_image_load_store, GL_AMD_performance_monitor, GL_EXT_separate_specular_color, GL_ARB_get_texture_sub_image, GL_ARB_pixel_buffer_object, GL_ARB_vertex_attrib_binding, GL_EXT_texture_mirror_clamp, GL_ARB_texture_mirrored_repeat, GL_EXT_texture_env_dot3, GL_ARB_map_buffer_alignment, GL_EXT_blend_subtract, GL_ANGLE_texture_compression_dxt3, GL_EXT_secondary_color, GL_EXT_blend_func_separate, GL_ARB_sparse_texture, GL_ARB_ES2_compatibility, GL_ARB_viewport_array, GL_SUN_multi_draw_arrays, GL_AMD_pinned_memory, GL_IBM_multimode_draw_arrays, GL_ARB_spirv_extensions, GL_ARB_map_buffer_range, GL_EXT_semaphore, GL_ARB_texture_compression_bptc, GL_MESA_shader_integer_functions, GL_ARB_shader_storage_buffer_object, GL_ARB_compressed_texture_pixel_storage, GL_EXT_draw_instanced, GL_ARB_texture_cube_map_array, GL_EXT_texture_sRGB_RG8, GL_ARB_window_pos, GL_ARB_shader_bit_encoding, GL_ARB_gl_spirv, GL_ARB_stencil_texturing, GL_ARB_shader_clock, GL_EXT_direct_state_access, GL_EXT_texture_lod_bias, GL_ARB_robustness, GL_AMD_vertex_shader_layer, GL_MESA_pack_invert, GL_ARB_draw_instanced, GL_ARB_fragment_program_shadow, GL_ARB_texture_buffer_range, GL_ARB_texture_buffer_object, GL_ARB_texture_buffer_object_rgb32, GL_ARB_shading_language_packing, GL_ARB_texture_non_power_of_two, GL_ARB_fragment_coord_conventions, GL_ARB_texture_border_clamp, GL_EXT_texture_sRGB_decode, GL_ARB_texture_mirror_clamp_to_edge, GL_EXT_vertex_array_bgra, GL_ARB_shader_atomic_counter_ops, GL_EXT_draw_buffers2, GL_ARB_gpu_shader_fp64, GL_AMD_shader_stencil_export, GL_EXT_texture_swizzle, GL_KHR_robustness, GL_ARB_ES3_compatibility, GL_ARB_debug_output, GL_SGIS_texture_lod, GL_NV_copy_depth_to_color, GL_EXT_packed_float, GL_AMD_texture_texture4, GL_EXT_depth_bounds_test, GL_EXT_bgra, GL_NV_primitive_restart, GL_ARB_seamless_cube_map, GL_EXT_texture_rectangle, GL_ARB_base_instance, GL_IBM_texture_mirrored_repeat, GL_NV_shader_atomic_int64, GL_EXT_demote_to_helper_invocation, GL_EXT_packed_pixels, GL_ARB_shader_objects, GL_ARB_shader_ballot, GL_ARB_texture_compression, GL_EXT_fog_coord, GL_ARB_texture_env_combine, GL_ARB_blend_func_extended, GL_ARB_enhanced_layouts, GL_AMD_draw_buffers_blend, GL_EXT_stencil_wrap, GL_EXT_texture_edge_clamp, GL_ARB_sparse_texture_clamp, GL_EXT_blend_minmax, GL_NV_blend_square, GL_ARB_derivative_control, GL_ARB_texture_cube_map, GL_ARB_clear_buffer_object, GL_EXT_texture_env_add, GL_EXT_texture_compression_latc, GL_NV_depth_clamp, GL_ARB_indirect_parameters, GL_ARB_texture_compression_rgtc, GL_KHR_context_flush_control, GL_KHR_texture_compression_astc_ldr, GL_ARB_texture_storage, GL_NV_alpha_to_coverage_dither_control, GL_ARB_depth_buffer_float, GL_ARB_shader_image_size, GL_ARB_provoking_vertex, GL_EXT_texture_shadow_lod, GL_ARB_framebuffer_sRGB, GL_ARB_shader_atomic_counters, GL_NV_texture_barrier, GL_ARB_sampler_objects, GL_NV_half_float, GL_AMD_query_buffer_object, GL_ARB_arrays_of_arrays, GL_AMD_framebuffer_multisample_advanced, GL_EXT_memory_object, GL_ARB_explicit_attrib_location, GL_MESA_window_pos, GL_ARB_vertex_array_object, GL_EXT_texture_sRGB, GL_ARB_transform_feedback2, GL_ARB_sparse_texture2, GL_OES_read_format, GL_S3_s3tc, GL_ARB_texture_storage_multisample, GL_ARB_shading_language_100, GL_EXT_framebuffer_object, GL_NV_conditional_render, GL_EXT_draw_range_elements, GL_ARB_vertex_buffer_object, GL_EXT_multi_draw_arrays, GL_ARB_explicit_uniform_location, GL_EXT_blend_equation_separate, GL_NV_light_max_exponent, GL_ARB_depth_texture, GL_ARB_vertex_program, GL_NV_texture_rectangle, GL_EXT_texture, GL_ARB_shader_draw_parameters, GL_ARB_clear_texture, GL_ARB_texture_swizzle, GL_ARB_buffer_storage, GL_INGR_blend_func_separate, GL_ARB_fragment_shader, GL_ARB_occlusion_query2, GL_ARB_draw_buffers, GL_ARB_draw_elements_base_vertex, GL_MESA_framebuffer_flip_y, GL_EXT_abgr, GL_ARB_shadow, GL_ARB_program_interface_query, GL_ARB_copy_buffer, GL_EXT_rescale_normal, GL_SGIS_generate_mipmap, GL_EXT_texture_compression_rgtc, GL_EXT_timer_query, GL_EXT_vertex_attrib_64bit, GL_ARB_seamless_cubemap_per_texture, GL_KHR_no_error, GL_AMD_multi_draw_indirect, GL_ARB_texture_float, GL_EXT_texture_sRGB_R8, GL_NV_compute_shader_derivatives, GL_ARB_robust_buffer_access_behavior, GL_EXT_texture_filter_anisotropic, GL_SGIS_texture_edge_clamp, GL_EXT_semaphore_fd, GL_EXT_vertex_array, GL_AMD_depth_clamp_separate, GL_ARB_texture_query_levels, GL_EXT_texture_compression_s3tc, GL_ARB_tessellation_shader, GL_ARB_shader_stencil_export, GL_ARB_shader_image_load_store, GL_EXT_texture_compression_dxt1, GL_ARB_texture_rgb10_a2ui, GL_EXT_provoking_vertex, GL_ARB_invalidate_subdata, GL_NV_texture_env_combine4, GL_ATI_blend_equation_separate, GL_ARB_vertex_type_2_10_10_10_rev, GL_ARB_direct_state_access, GL_ATI_texture_mirror_once, GL_ARB_sparse_buffer, GL_ARB_shader_subroutine, GL_ARB_transform_feedback3, GL_ATI_texture_compression_3dc, GL_ARB_post_depth_coverage, GL_EXT_shader_samples_identical, GL_ARB_instanced_arrays, GL_ARB_occlusion_query, GL_KHR_texture_compression_astc_sliced_3d, GL_ARB_separate_shader_objects, GL_EXT_texture_cube_map, GL_EXT_compiled_vertex_array, GL_EXT_texture_env_combine, GL_ARB_internalformat_query, GL_EXT_gpu_shader4, GL_ARB_bindless_texture, GL_ARB_gpu_shader_int64, GL_ARB_multi_bind, GL_ARB_texture_env_add, GL_IBM_rasterpos_clip, GL_EXT_pixel_buffer_object, GL_ARB_texture_barrier, GL_ARB_transform_feedback_instanced, GL_APPLE_packed_pixels, GL_ARB_vertex_attrib_64bit, GL_NV_texgen_reflection, GL_ARB_multitexture, GL_ARB_multisample, GL_ATI_separate_stencil, GL_EXT_texture_object, GL_ARB_pipeline_statistics_query, GL_ARB_color_buffer_float, GL_ARB_multi_draw_indirect, GL_ARB_framebuffer_no_attachments, GL_ARB_compute_shader, GL_ARB_transpose_matrix, GL_ARB_vertex_array_bgra, GL_NVX_gpu_memory_info, GL_ARB_shader_texture_lod, GL_ARB_fragment_program, GL_AMD_vertex_shader_viewport_index, GL_EXT_texture_array, GL_ARB_draw_indirect, GL_ARB_half_float_vertex, GL_EXT_blend_color, GL_ARB_texture_filter_anisotropic, GL_ANGLE_texture_compression_dxt5, GL_ARB_texture_env_crossbar, GL_EXT_window_rectangles, GL_ARB_clip_control, GL_ARB_vertex_shader, GL_ARB_compute_variable_group_size
[2023.05.29 10:47:52] OpenGL: QOpenGLContext created, version: 4.4.
[2023.05.29 10:47:52] App Info: reading accounts info...
[2023.05.29 10:47:52] App Info: reading encrypted info...
[2023.05.29 10:47:52] App Info: reading map...
[2023.05.29 10:47:52] App Info: reading encrypted map...
[2023.05.29 10:47:52] App Info: reading encrypted user settings...
[2023.05.29 10:47:52] App Info: encrypted user settings read.
[2023.05.29 10:47:52] App Info: reading encrypted mtp data...
[2023.05.29 10:47:52] MTP Info: read keys, current: 5, to destroy: 0
[2023.05.29 10:47:52] Map read time: 2
[2023.05.29 10:47:52] App Info: reading encrypted mtp config...
[2023.05.29 10:47:53] Export Info: Destroy top bar by controller removal.
[2023.05.29 10:47:54] OpenGL: [TRUE] (Window)
[2023.05.29 10:47:54] Notification daemon product name: gnome-shell
[2023.05.29 10:47:54] Notification daemon vendor name: GNOME
[2023.05.29 10:47:54] Notification daemon version: 42.5
[2023.05.29 10:47:54] Notification daemon specification version: 1.2
[2023.05.29 10:47:54] Notification daemon capabilities: actions, body, body-markup, icon-static, persistence, sound
[2023.05.29 10:47:54] API Warning: not loaded minimal channel applied.
[2023.05.29 10:47:55] API Warning: not loaded minimal channel applied.
[2023.05.29 10:47:55] API Warning: not loaded minimal channel applied.
[2023.05.29 10:47:55] API Warning: not loaded minimal channel applied.
[2023.05.29 10:47:55] API Warning: not loaded minimal channel applied.
[2023.05.29 10:47:57] API Warning: not loaded minimal channel applied.
[2023.05.29 10:51:44] Audio Info: recreating audio device and reattaching the tracks
[2023.05.29 10:51:46] Audio Info: Closing audio playback device.
```

**Comment by ilya-fedin** [comment]
I can't reproduce, provide a screen record of the issue please

**Comment by LDVSOFT** [comment]
How could I share recording privately? It obviously would expose my chats.

**Comment by Aokromes** [comment]
> How could I share recording privately? It obviously would expose my chats.

you can create a folder with for example telegram beta chat and screenrecord only that folder.

**Comment by LDVSOFT** [comment]
Ok, GitHub limits file to 10Mb, so you can look [here](https://ldvsoft.net/static/tdesktop%20lags.mp4).

**Comment by ilya-fedin** [comment]
Can't reproduce

**Comment by ilya-fedin** [comment]
Well, GNOME always makes everything for Qt apps to work badly, I'm not surprised. I don't think it will ever be possible to get a good UX with tdesktop on GNOME desktop, due to all these silent cross-toolkit wars.

**Comment by PoSayDone** [comment]
I have the same issue with Telegram from Arch Linux repo and with Telegram from official website. I am using Hyprland, not gnome and have massive stutters and lags, but don't have them on X11

## Specs
* telegram 4.8.3
* ArchLinux
* Intel GPU
* Hyprland

**Comment by ilya-fedin** [comment]
Will it lag on Plasma Wayland for you? I wonder whether it's DE-dependent, hardware-dependent or both... I have AMD hardware and Plasma Wayland, it works pretty fine...

**Comment by PoSayDone** [comment]
> Will it lag on Plasma Wayland for you? I wonder whether it's DE-dependent, hardware-dependent or both... I have AMD hardware and Plasma Wayland, it works pretty fine...

Seems that I don't have such issue in plasma

**Comment by ilya-fedin** [comment]
Does that mean Qt has bad compatibility with everything except of Plasma? Is it reproducible with Weston?

**Comment by PoSayDone** [comment]
> Does that mean Qt has bad compatibility with everything except of Plasma? Is it reproducible with Weston?

No, don't have such issue in Weston.
Also checked gnome and the issue is reproducible in it.

**Comment by ilya-fedin** [comment]
I wonder what GNOME and Hyprland do that Qt operates slowly? :thinking:

**Comment by PoSayDone** [comment]
> I wonder what GNOME and Hyprland do that Qt operates slowly? thinking

Also, other QT apps don't have such issue, for example qbittorrent, OBS and WPS Office

**Comment by ilya-fedin** [comment]
Are you sure they're based on Qt 6 and use OpenGL as well as tdesktop?

**Comment by vitordelucca** [comment]
I have the exactly same problem. It also weirdly consume way, way more RAM than necessary.

**Comment by ilya-fedin** [comment]
@vitordelucca how much?

**Comment by vitordelucca** [comment]
> @vitordelucca how much?

Easily 500mb... I was used to tdesktop on Windows, when after hours and hours of use, was always less than 100mb.

Anyway, just disabled "OpenGL for media" on telegram settings and perf animations improved under wayland.

**Comment by ilya-fedin** [comment]
> Easily 500mb... I was used to tdesktop on Windows, when after hours and hours of use, was always less than 100mb.

Well, it starts from 400MB on Linux so it's normal... I guess that's some difference in how the systems work.

> I wonder if there is some flag to enable xwayland on tdesktop to see if perf improves?

QT_QPA_PLATFORM=xcb

About OpenGL for media (which was removed from the comment), disabling it forces SHM flush which is reported to be working poorly on GNOME, leading to graphics artifacts on some GNOME systems. Apparently OpenGL also works poorly on some GNOME systems, which means both Qt's flush mechanism are broken on GNOME's Wayland implementations in various ways... The slow performance is probably better than graphics artifacts so I guess switching the default is not a good idea. And I have no idea how to get this fixed given that the one who can reproduce (and I can't) should create a minimal reproducible example for Qt developers to accept a bug report.

**Comment by vitordelucca** [comment]
> QT_QPA_PLATFORM=xcb

How can I set it?

**Comment by ilya-fedin** [comment]
`QT_QPA_PLATFORM=xcb ./Telegram`
Telegram shouldn't running in background or it would just send a message to an already running instance.

**Comment by vitordelucca** [comment]
> QT_QPA_PLATFORM=xcb ./Telegram

Indeed this way is much faster.... I'm wondering if this could be put into experimental settings for those who might have problems turn on easily?

**Comment by ilya-fedin** [comment]
@vitordelucca
1. I'm not sure it's worth doing
2. Environment variables leak into child processes so if you open some downloaded file in tdesktop that involves launching an external application and that application is a Qt one, it will inherit tdesktop's environment including this variable. This usually leads to bug reports like #10487.

**Comment by vitordelucca** [comment]
Found a way to disable Wayland for QT Flatpak apps: Using Flatseal, just unselect Wayland, and then Telegram will run as X11 (so in this case, x)

**Comment by ZerdoX-x** [comment]
Hi gnome users :) As Ilya mentioned
> Well, GNOME always makes everything for Qt apps to work badly, I'm not surprised

Wayland gnome implements protocols very poorly in contrast with other DEs and WMs, do not use it if you want good wayland setup and environment, you may encounter issues in many apps. Especially if its combo with nvidia gpu, that's the worst setup you can do :) I am sure that telegram-desktop can do nothing about issues on gnome's wayland. It just uses qt and opengl, that's it.

---

I was using telegram-desktop for years. My wayland experience with it included 2 setups:
 - huawei laptop with amd graphics, void linux, river wm
 - apple air m1 laptop, gentoo linux, hyprland wm
 
 In both cases it's high res and scaling, do not experience any issues. On my second setup it's even pure wayland, I don't have X installed.
 
 Current setup:
  - 2560x1600@60 monitor with 1.5 scaling
  - `net-im/telegram-desktop-4.10.3-r1` compiled with source with `wayland dbus fonts screencast -X -enchant -qt6-imageformats -webkit` compilation flags
  - qt5, did not have time yet figuring out how to migrate to qt6
  - hyprland 0.30.0. wlroots 0.16.2
  - latest asahi video drivers
  
 I am not sure if it's 4k or not, but should be pretty close. I also use blur behind telegram window and lot's of opacity in my theme, it does not give me any performance issues.
 
 ---
 
 > I wonder what GNOME and Hyprland do that Qt operates slowly? 🤔
 
 I am sure Hyprland does nothing with Qt or any other frameworks, it's just a simple wlroots based WM.
 
 ---
 
**So current official distribution uses qt6?** I think I'll find some time to compile it under qt6 and give feedback here.
I was just looking through some open issues and stumbled across this one, so why not to help you testing this :)

---

BTW, Ilya, is it possible to hide these icons on telegram window?

<details>
<summary>Screenshot</summary>

![image](https://github.com/telegramdesktop/tdesktop/assets/49815452/59bdc022-7ef7-4c9f-a27d-c5e481f9bd21)

</details>

I tried setting env variable `QT_WAYLAND_DISABLE_WINDOWDECORATION` to `1` but tg window just ignores it. Please let me know if there is a quick way to hide them :)

**Comment by ilya-fedin** [comment]
Enable Qt window frame? Telegram's window frame follows GTK's window frane layout via portal.

**Comment by ilya-fedin** [comment]
> I am sure Hyprland does nothing with Qt or any other frameworks, it's just a simple wlroots based WM.

Well, almost every Qt update leads to bug reports from wlroots and GNOME users. These issues are frequently reproducible even with weston but testing under kwin never shows issues. So there's definitely some difference between kwin and every other compositor that makes Qt utterly buggy on everything but kwin.

Your good experience might because you're stumbled upon Qt 5 that had no minor updates for years...

**Comment by ZerdoX-x** [comment]
> Your good experience might be because you're stumbled upon Qt 5 that had no minor updates for years...

Just unmasked qt6 use flag on my arm64 gentoo and recompiled all packages that can use it. Qt6 is still masked in gentoo for all architectures except amd64 unstable as it's considered not ready for migration from 5 version for all users, and most packages still use qt5 afaik.

_So now some of my qt apps like obs and telegram use qt6. Feels like transparent themes in telegram-desktop work better then before, nice :D_ And again I did not stumbled upon any performance issues.

May this be related that guys using not the latest qt dependencies? I have ton of them on gentoo, and sometimes after @world upgrade I can stumble upon some issues like _"freezing when reading images"_, I came across some of issues like this in this repo, but I don't really care about them as they can appear when some little dependency was updated like `dev-qt/qtsvg`, etc.. May this be related to ubuntu/debian having legacy libs/dependencies/packages and all that? I update my packages once a week and sometimes it can cause issues in telegram, even on updates that came to qt5 components, and even if tg itself wasn't upgraded. Though I can't remember huge performance loss as being reported in this issue.

Maybe we can list all telegram dependencies and require at least somewhat up to date?
Here are all qt dependencies from telegram-desktop I have:
```
dev-qt/qtcore-5.15.11-r1
dev-qt/qtgui-5.15.11
dev-qt/qtnetwork-5.15.11
dev-qt/qtwidgets-5.15.11
kde-frameworks/kcoreaddons-5.111.0
dev-qt/qtbase-6.6.0-r1
dev-qt/qtimageformats-6.6.0
dev-qt/qtsvg-6.6.0-r1
dev-qt/qtdeclarative-6.6.0
dev-qt/qtwayland-6.6.0
```

And while typing this I found that when upgraded to qt6 my WM workspace gets switched to one where telegram-desktop is opened, so it just requires focus every time I get a message, notification. This is really awful :0 Is there a way to disable this? Is this considered an expected behavior? I know I can disable this globally in WM, but sometimes it's useful, in some apps, but not every time you receive a notification..

---

> Enable Qt window frame? Telegram's window frame follows GTK's window frame layout via portal.

~I remember there was a setting like this in telegram-desktop, but I could not find it anymore, even in dev settings~
Found it, thank you! Maybe it appeared when I migrated to qt6, not sure why I was not able to find it.

**Comment by ilya-fedin** [comment]
> May this be related that guys using not the latest qt dependencies?

I'd rather assume it's related to GL drivers and Qt<->GL (mesa)<->compositor communication

**Comment by ilya-fedin** [comment]
> so it just requires focus every time I get a message, notification

Apparently that's how your compositor reacts to QWindow::alert? It blinks in taskbar on KDE.

**Comment by ZerdoX-x** [comment]
> I'd rather assume it's related to GL drivers and Qt<->GL (mesa)<->compositor communication

Hm, this issue, maybe. I was just thinking about my cases when I saw freezes after upgrading some little qt deps.

---

> Apparently that's how your compositor reacts to QWindow::alert?

Oookayy.. Seems like this is intended in hyprland. There are some issues and discussion related to this:
1. https://github.com/hyprwm/Hyprland/issues/2732 
2. https://github.com/hyprwm/Hyprland/issues/1350

Hyprland creator quote from https://github.com/hyprwm/Hyprland/issues/1350#issuecomment-1377065229:
> the reason I keep it off and generally not work on it is that it was causing me more issues than switching workspaces to firefox. Maybe it's my shady impl, maybe not, but some apps would repeatedly spam activation a kjhillion times a second (e.g. obs)

I'll post your message above to one of the threads and we'll discuss this behavior, maybe something wrong with implementation of focus request. Thanks.

---

> It blinks in taskbar/dock on KDE

I don't use any taskbars or docks :)

---

Anyways I don't want this issue gets flooded by off topic messages, so thank you for all that info, and I gave you as much info as I can. I am not subscribing to the thread, so do not hesitate pinging me if you want me to test anything. The problem here is that I am using asahi video drivers which is crazy and only few people here use it :)

**Comment by vitordelucca** [comment]
> Hi gnome users :) As Ilya mentioned
> 
> > Well, GNOME always makes everything for Qt apps to work badly, I'm not surprised
> 
> Wayland gnome implements protocols very poorly in contrast with other DEs and WMs, do not use it if you want good wayland setup and environment, you may encounter issues in many apps. Especially if its combo with nvidia gpu, that's the worst setup you can do :) I am sure that telegram-desktop can do nothing about issues on gnome's wayland. It just uses qt and opengl, that's it.
> 
> I was using telegram-desktop for years. My wayland experience with it included 2 setups:
> 
>     * huawei laptop with amd graphics, void linux, river wm
> 
>     * apple air m1 laptop, gentoo linux, hyprland wm
> 
> 
> In both cases it's high res and scaling, do not experience any issues. On my second setup it's even pure wayland, I don't have X installed.
> 
> Current setup:
> 
>     * 2560x1600@60 monitor with 1.5 scaling
> 
>     * `net-im/telegram-desktop-4.10.3-r1` compiled with source with `wayland dbus fonts screencast -X -enchant -qt6-imageformats -webkit` compilation flags
> 
>     * qt5, did not have time yet figuring out how to migrate to qt6
> 
>     * hyprland 0.30.0. wlroots 0.16.2
> 
>     * latest asahi video drivers
> 
> 
> I am not sure if it's 4k or not, but should be pretty close. I also use blur behind telegram window and lot's of opacity in my theme, it does not give me any performance issues.
> 
> > I wonder what GNOME and Hyprland do that Qt operates slowly? 🤔
> 
> I am sure Hyprland does nothing with Qt or any other frameworks, it's just a simple wlroots based WM.
> 
> **So current official distribution uses qt6?** I think I'll find some time to compile it under qt6 and give feedback here. I was just looking through some open issues and stumbled across this one, so why not to help you testing this :)
> 
> BTW, Ilya, is it possible to hide these icons on telegram window?
> Screenshot
> 
> I tried setting env variable `QT_WAYLAND_DISABLE_WINDOWDECORATION` to `1` but tg window just ignores it. Please let me know if there is a quick way to hide them :)

I don't think it's related to Gnome. I'm now running EndeavourOS with KDE, and I still have bad Telegram perf under Wayland. When I run Telegram as xwayland, it runs much better.

I didn't notice this problem on any other QT app (after all, I'm on KDE, everything is QT!). Even SDDM I'm running under wayland and is working fine. 

Only Tegram have this issue.

**Comment by K900** [comment]
Can reproduce on Plasma Wayland too.

**Comment by ilya-fedin** [comment]
It's quite sad that no one who reproduces manage to debug the issue...

**Comment by ilya-fedin** [comment]
Can everyone tell their screen resolution Telegram is slow on, screen scale Telegram is slow on, the number of screens, Telegram window size (mentioning whether it's physical size or divided by scale), whether the issue disappears after shrinking the window size?

**Comment by PoSayDone** [comment]
> Can everyone tell their screen resolution & screen scale of all screens mentioning Telegram is slow on which of them, Telegram window size (mentioning whether it's physical size or divided by scale), whether the issue disappears after shrinking the window size?

2880x1800
scale 1.8

**Comment by pranayadmn** [comment]
> Can everyone tell their screen resolution & screen scale of all screens mentioning Telegram is slow on which of them, Telegram window size (mentioning whether it's physical size or divided by scale), whether the issue disappears after shrinking the window size?

1920x1080, TDesktop is laggy for me when using a screen scale of 1.25 but seems to be fine on the default scale.
Also, the issue is indeed fixed when the window is shrunk from the maximum size.

**Comment by ilya-fedin** [comment]
@pranayadmn what's the CPU and GPU in your case? As I also have 1920x1080 with 1.25 and can't reproduce :(

**Comment by andresilva** [comment]
3840x2160 resolution with 1.5 scaling (using Hyprland). Can confirm that the issue seems to be gone if scaling is disabled or the window is set to a smaller size.

> The "Enable precise High DPI scaling" option in Experimental settings also helps to some extent.

Indeed it seems to make it better, although it introduces some other weird behavior (icons and emojis look a bit pixelated).

**Comment by pranayadmn** [comment]
> @pranayadmn what's the CPU and GPU in your case? As I also have 1920x1080 with 1.25 and can't reproduce :(

i3-8100 with an NVIDIA GT 1030 w/ 545.29.02 drivers, running on GNOME.

**Comment by ilya-fedin** [comment]
@pranayadmn are you sure it's NVIDIA sometimes people with dual GPU system say NVIDIA but NVIDIA is not used on such systems for anything but games?

**Comment by pranayadmn** [comment]
> @pranayadmn are you sure it's NVIDIA? sometimes people with dual GPU system say NVIDIA but NVIDIA is not used on such systems for anything but games?

I can confirm the GPU in use is indeed the GT 1030 via resource monitoring tools which show its utilisation and the iGPU utilisation to be constantly 0%. Although I don't think it matters in my case for a desktop since the display is directly plugged in to the external GPU.

**Comment by ilya-fedin** [comment]
@pranayadmn can you provide tdesktop's log.txt?

**Comment by pranayadmn** [comment]
> @pranayadmn can you provide tdesktop's log.txt?

[log.txt](https://github.com/telegramdesktop/tdesktop/files/13336167/log.txt)
Here you go!

**Comment by ilya-fedin** [comment]
Will it still happen if you click reset the experimental options?

**Comment by pranayadmn** [comment]
> @pranayadmn Will it still happen if you click reset the experimental options and restart the client?

It might have been left accidentally enabled by me when I was trying to find what could possibly fix the issue. Yes, it does happen even when the experimental settings are disabled.

**Comment by ilya-fedin** [comment]
Does it happen with animations disabled and when no moving parts are on screen? Maybe the CPU just can't keep up with drawing the animations/moving content fast enough (yeah, tdesktop main window is painted using CPU as any QtWidgets application) on a scaled surface (which is way bigger as it's painted at the next integer factor which is 2x)... If it is then apparently it's not possible to fix without fixing #1121 first, which in turn is not possible to fix without full rewrite of application 😭

**Comment by pranayadmn** [comment]
> Does it happen with animations disabled and when no moving parts are on screen? Maybe the CPU just can't keep up with drawing the animations/moving content fast enough (yeah, tdesktop main window is painted using CPU as any QtWidgets application) on a scaled surface (which is way bigger as it's painted at the next integer factor which is 2x)... If it is then apparently it's not possible to fix without fixing #1121 first, which in turn is not possible to fix without full rewrite of application 😭
> 
> I guess no other QtWidgets application has animations nor such moving content which would explain why it's not possible to reproduce with any other Qt application.

It is indeed snappier when interface animations are disabled, but, isn't that kinda expected? 😅
I could consider switching to Xorg but the lack of fractional scaling support makes the display elements miserably small, while Xorg apps under Wayland look awful with fractional scaling :/

**Comment by ilya-fedin** [comment]
Yeah, as you can see, tdesktop is a legacy application regarding scaling. This is unlikely to improve so you can switch to the Electron version from the official Web A repo if that's something you really need. The best thing could be done in tsesktop is to patch Qt in official builds to always work at 1x scale but I guess this will make things worse as the compositor will upscale then and tdesktop window will be blurry.

**Comment by ilya-fedin** [comment]
Can everyone confirm that this happens only with low-end processors having 4 threads and doesn't happen with processor having 8 threads and more?

**Comment by K900** [comment]
It is happening for me on a Ryzen 9 6900HS, with 8 cores + SMT.

**Comment by andresilva** [comment]
Happening for me on a Ryzen 9 5950X with 16 cores.

**Comment by K900** [comment]
I am confident it's not (just) a performance issue, because the exact same window has no trouble rendering animations on X11.

**Comment by ilya-fedin** [comment]
X11 renders 1:1. While on Wayland with fractional scaling applications having no support for fractional scale factors (like tdesktop) get fake 2x scaling factor and the compositor downscales the buffer in raster.

**Comment by ilya-fedin** [comment]
so if you have 4K@1.25x on X11, applications renders in 4K@1.25x. On Wayland it would be around 16K@2x and then composited down to 4K by the compositor.

**Comment by ilya-fedin** [comment]
tdesktop's bad HiDPI support is not seen on X11 only because it's as bad as tdesktop itself. Windows users have a lot of woes with tdesktop's HiDPI which could be seen by the complains in #1121. E.g. tdesktop doesn't support different scale factors on different screens (yet Qt could force-rescale with quality loss, it's especially visible with the precise High DPI scaling experimental option). It also doesn't support fractional device pixel ratios as they appeared long after 2013 when tdesktop was developed (yet other platforms don't have scale confirmation dance like Wayland so no upscaling/downscaling happens and tdesktop can scale up the fractional part by itself without Qt knowing that).

**Comment by andresilva** [comment]
```
❯ cat .local/share/TelegramDesktop/DebugLogs/log_18_30.txt | grep -i "Window Pos"
[18:31:54.574 00-0000672] Window Pos: Initializing first 0, 0, 2550, 1388 (scale 125%, maximized 0) (core_settings.cpp : 25)
[18:31:54.574 00-0000673] Window Pos: Screen found, screen geometry: 0, 0, 2560, 1440, available: 0, 0, 2560, 1440 (main_window.cpp : 25)
[18:31:54.574 00-0000674] Window Pos: Window frame margins: 0, 0, 0, 0, available space for inner geometry: 0, 0, 2560, 1440 (main_window.cpp : 25)
[18:31:54.574 00-0000675] Window Pos: Resulting geometry is 0, 0, 2550, 1388 (main_window.cpp : 25)
[18:31:54.574 00-0000676] Window Pos: Setting first 0, 0, 2550, 1388 (main_window.cpp : 25)
[18:31:55.756 00-0001570] Window Pos: Saving non-maximized position: 0, 0, 2550, 1388 (main_window.cpp : 25)
[18:31:55.756 00-0001571] Window Pos: Screen found, geometry: 0, 0, 2560, 1440 (main_window.cpp : 25)
```

As I said above my monitor resolution is 3840x2160 with 1.5x scaling. Interface scale in telegram options is set to 250%.

**Comment by ilya-fedin** [comment]
@andresilva actual buffer size should be 5120x2776 for you. So it renders in 5K with 2.5x scaling...

**Comment by andresilva** [comment]
Guess I'll stick to using the app with `QT_QPA_PLATFORM=xcb`. With the in-app scaling it ends up working OK (unlike some other apps running on xwayland).

**Comment by PoSayDone** [comment]
> Can everyone confirm that this happens only with low-end processors having 4 threads and doesn't happen with processors having 8 threads and more?

no, happening for me with i7 1360p, 12th cores 16th threads.

**Comment by ilya-fedin** [comment]
@PoSayDone yet it becomes better if you shrink the window?

**Comment by PoSayDone** [comment]
> @PoSayDone yet it becomes better if you shrink the window?

yes it is

**Comment by ilya-fedin** [comment]
A mystery. 2880x1800 with 1.8 scale should be 3200x2000 which is near to 1920x1080 with 1.25 scale (3072x1728), I'd expect such a processor to keep up with that...

**Comment by ilya-fedin** [comment]
Apparently we're where we were: this could be debugged only by people who can reproduce and sadly I have no idea what to suggest for debugging.

**Comment by ilya-fedin** [comment]
Ah, last idea: D-Bus was a source of slowdowns in the past, does setting `DBUS_SESSION_BUS_ADDRESS` to a random value (to prevent D-Bus connection) before launching tdesktop helps?

**Comment by pranayadmn** [comment]
> Ah, two last ideas:
> 
> 1. D-Bus was a source of slowdowns in the past, does setting `DBUS_SESSION_BUS_ADDRESS` to a random value (to prevent D-Bus connection) before launching tdesktop help?
> 2. Tray icon was a source of slowdowns in the past as well, does disabling it help?

Can confirm that:
1. When using `DBUS_SESSION_BUS_ADDRESS=0 telegram-desktop -debug` the logs do say that dbus session is not present but it does not improve the interface performance either.
2. This does not work as well.

**Comment by ilya-fedin** [comment]
> The best thing could be done in tdesktop is to patch Qt in official builds to always work at 1x scale

Looks like I was wrong on this one, I managed to patch Qt to follow the rounding policy while having fractional-scale-v1 support active. tdesktop no longer autodetects fractional scale with such a patch but renders 1:1. If I manage to activate experimental Qt-level downscaling via the precise experimental option and a way to detect whether the patch is applied perhaps this could be shipped.

**Comment by ilya-fedin** [comment]
Here's a binary built with Qt patches I made. Please test whether the issue happens with both disabled and enabled precise High-DPI option.
https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing

**Comment by pranayadmn** [comment]
> Here's a binary built with Qt patches I made. Please test whether the issue happens with both disabled and enabled precise High-DPI option. https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing
> 
> I suggest to create a TelegramForcePortable directory near the binary so it doesn't use the data of already existing tdesktop installation.

It now lags with "Enable precise High DPI scaling" option instead and also introduces some glitches in sub-menus with that option. When it's disabled, it runs as if it was running with the default screen scale instead of 1.25.

**Comment by ilya-fedin** [comment]
Yeah, it can't autodetect your scale anymore with this patch, please choose the right scale manually in application settings

**Comment by ilya-fedin** [comment]
Please share a screenshot of the sub-menu glitches

**Comment by pranayadmn** [comment]
> Yeah, it can't autodetect your scale anymore with this patch, please choose the right scale manually in application settings

Oh okay, so it was supposed to be like that.

**Comment by pranayadmn** [comment]
> Please share a screenshot of the sub-menu glitches

![Screenshot from 2023-11-14 21-03-51](https://github.com/telegramdesktop/tdesktop/assets/61061454/1d08b81b-111f-4f22-99f3-6812ba8572a6)

**Comment by pranayadmn** [comment]
> > Here's a binary built with Qt patches I made. Please test whether the issue happens with both disabled and enabled precise High-DPI option. https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing
> > I suggest to create a TelegramForcePortable directory near the binary so it doesn't use the data of already existing tdesktop installation.
> 
> It now lags with "Enable precise High DPI scaling" option instead and also introduces some glitches in sub-menus with that option. When it's disabled, it runs as if it was running with the default screen scale instead of 1.25.

Apparently, it's still a bit random. With the precise DPI option disabled, it seems that new panels still take some time to be drawn, but once it's done I can navigate through them smoothly. Here's a screen recording clip that demonstrates it:
[Screencast from 2023-11-14 21-10-58.webm](https://github.com/telegramdesktop/tdesktop/assets/61061454/9a20e100-fa9c-4689-bc80-ec4ee59e0234)

**Comment by ilya-fedin** [comment]
> Also notice the random artifacts on the menus even when the precise DPI option is disabled.

It's apparently a bug in fractional-scale-v1 implementation. I have them as well, if you're using kwin too then idk whose, if other compositor then apparently it's a Qt bug.

**Comment by ilya-fedin** [comment]
While uploading the binary I found that i have a build without jemalloc. Can you test it as well (maybe jemalloc makes memory (de)allocations slowly and slows down everything, who knows)? https://drive.google.com/file/d/1QcWyPXPN8RVpki2qjr9Ranz71ycvH2yr/view?usp=sharing

Make extra sure to have TelegramForcePortable folder with this one as it's an old version and old versions clean up data directory if they find the data is from a newer version.

**Comment by pranayadmn** [comment]
> While uploading the binary I found that i have a build without jemalloc. Can you test it as well (maybe jemalloc makes memory (de)allocations slowly and slows down everything, who knows)? https://drive.google.com/file/d/1QcWyPXPN8RVpki2qjr9Ranz71ycvH2yr/view?usp=sharing
> 
> Make extra sure to have TelegramForcePortable folder with this one as it's an old version and old versions clean up data directory if they find the data is from a newer version.

Still the same I'm afraid :(

**Comment by ilya-fedin** [comment]
One person in this thread has mentioned (but later removed from the message)
> Anyway, just disabled "OpenGL for media" on telegram settings and perf animations improved under wayland.

I wonder whether anyone else checked that?

**Comment by ilya-fedin** [comment]
Could it be related to screen refresh rate? :thinking:

**Comment by pranayadmn** [comment]
> Could it be related to screen refresh rate? 🤔

Mine's 60hz so I'm not quite sure.

**Comment by ilya-fedin** [comment]
Yeah, I have 60 Hz too, apparently it's not

**Comment by ilya-fedin** [comment]
> With the precise DPI option disabled, it seems that new panels still take some time to be drawn, but once it's done I can navigate through them smoothly.

Is this one any better (including sub-menu glitches)?
https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing

**Comment by PoSayDone** [comment]
> > With the precise DPI option disabled, it seems that new panels still take some time to be drawn, but once it's done I can navigate through them smoothly.
> 
> Is this one any better (including sub-menu glitches)? https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing

Not for me, still lagging on normal window size. Also got strange orange stripes all over rendered elements.
![image](https://github.com/telegramdesktop/tdesktop/assets/29358657/ad3e2e5a-0012-497d-baf7-f84981847a2a)

**Comment by pranayadmn** [comment]
> > With the precise DPI option disabled, it seems that new panels still take some time to be drawn, but once it's done I can navigate through them smoothly.
> 
> Is this one any better (including sub-menu glitches)? https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing

Now those lags are present no matter if "OpenGL for media" option is disabled or enabled. Also nope the right-click menu is still enlarged like before with precise DPI scaling.

**Comment by ilya-fedin** [comment]
Without the precise option enabled? That would be surprising as I changed QWaylandScreen to tell devicePixelRatio = 1 so it should render at 2x lesser size.

**Comment by pranayadmn** [comment]
> Without the precise option enabled? That would be surprising as I changed QWaylandScreen to tell devicePixelRatio = 1 so it should render at 2x lesser size.

Yep, it lags with both precise DPI scaling disabled or enabled in that build (only the panel drawing animation lags).

**Comment by ilya-fedin** [comment]
Ah, you mean it's the same as it was with the previous build?

**Comment by pranayadmn** [comment]
> Here's a binary built with Qt patches I made. Please test whether the issue happens with both disabled and enabled precise High-DPI option. https://drive.google.com/file/d/1JL5OhXJkh7mHurTPXtpaDyKEQc0r1Dl4/view?usp=sharing
> 
> I suggest to create a TelegramForcePortable directory near the binary so it doesn't use the data of already existing tdesktop installation.

If you mean this one then yes.

**Comment by ilya-fedin** [comment]
Ok, perhaps this doesn't really help, apparently the larger rendering surface is not the cause. I'm completely out of ideas though.

**Comment by gamelaster** [comment]
Strange, I began to have same issue, disabling animations solved it, but it started to happen only few days ago. I always used latest version from AUR.

**Comment by ilya-fedin** [comment]
Is it any better with 4.13?

**Comment by vitordelucca** [comment]
> Is it any better with 4.13?

Same thing here, sadly.

**Comment by ilya-fedin** [comment]
Has anyone tried to debug it the other way: by finding first bad version?

**Comment by ilya-fedin** [comment]
Was reading [the wayland-proxy news on Phoronix](https://www.phoronix.com/news/Wayland-Proxy-Firefox) and a new hypothesis has come to my mind: Qt has a separate thread for Wayland event loop since 6.0, that made it invulnerable for crashes due to hi-res mouses but I guess they can still flood the application with events to the point its performance is bad. Maybe that's what happens here?

**Comment by lilydjwg** [comment]
I have a normal mouse and can't reproduce the issue reported by Firefox users.

> Has anyone tried to debug it the other way: by finding first bad version?

But downgrading Telegram will cause a re-login. I'm afraid that repeated re-login may cause my account to be banned.

**Comment by ilya-fedin** [comment]
> But downgrading Telegram will cause a re-login. I'm afraid that repeated re-login may cause my account to be banned.

Start with some very old version? Like a year old, if it's broken, then two years old. And go from old to new ones. You can also create `TelegramForcePortable` folder near the Telegram binary so the main data folder is not affected.

**Comment by ilya-fedin** [comment]
Too old versions may run via Xwayland by default and it's a good idea to run with QT_QPA_PLATFORM=wayland to ensure native Wayland backend is tested

**Comment by ilya-fedin** [comment]
Is it any better with 4.14.3?

**Comment by lilydjwg** [comment]
> Is it any better with 4.14.3?

No.

I did a bisect (with btrfs snapshots so I could avoid repeated logins). Tested with maximized window and animated sticker panel open, the result is 4.7.1 is good (partial damage & power 57~65W) and 4.8.0 is bad (whole damage & power>=80).

**Comment by ilya-fedin** [comment]
> No.

With both OpenGL enabled and disabled?

**Comment by ilya-fedin** [comment]
> the result is 4.7.1 is good

With `QT_QPA_PLATFORM=wayland` or without? As far as I can see 4.8.0 is the first version removing GNOME from the native Wayland blacklist so it's possible it's just running via Xwayland.

**Comment by lilydjwg** [comment]
> With both OpenGL enabled and disabled?

I'm not sure which setting you mean. I didn't change anything related, starting with a fresh `-workdir`.

> With `QT_QPA_PLATFORM=wayland` or without?

With. I tested that the oldest version I tested with was running under native Wayland. Also I'm using Wayfire, not GNOME.

**Comment by ilya-fedin** [comment]
> I'm not sure which setting you mean.

I mean try to disable OpenGL in settings -> advanced

**Comment by lilydjwg** [comment]
It was enabled during the tests.

I tried to disable it just now and things returned to <=4.7.1 state.

**Comment by ilya-fedin** [comment]
I wonder why didn't it help to @pranayadmn

**Comment by ilya-fedin** [comment]
Is it the same with this version?
```
flatpak install --user https://dl.flathub.org/build-repo/74523/org.telegram.desktop.flatpakref
```

**Comment by gamelaster** [comment]
I tested 4.14.3, I assured that it is running in Wayland mode, and after enabling interface animations, Telegram became sluggish again, and when I turned off internet access, the reconnect animation hogged 10% - 30% CPU usage.

**Comment by ilya-fedin** [comment]
> the reconnect animation hogged 10% - 30% CPU usage.

This one I believe is cross-platfform and sadly expected...

**Comment by ilya-fedin** [comment]
I wonder whether the flatpak test build or snap edge is any better... There perhaps won't be a stable version for a month to test with the binary.

**Comment by lilydjwg** [comment]
> Is it the same with this version?
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/74523/org.telegram.desktop.flatpakref
> ```

Yes, it's the same.

**Comment by ilya-fedin** [comment]
Wow, I patched Qt to get rid of unneeded buffer copies so it's really weird

**Comment by lilydjwg** [comment]
Maybe it got slightly better but too slightly compared to OpenGL disabled?

**Comment by ilya-fedin** [comment]
idk, it's ok for me with both opengl enabled and disabled. Other people in the issue were saying that disabling opengl doesn't help so maybe there are multiple issues.

**Comment by lilydjwg** [comment]
Oh my mistake. The original one with OpenGL is 85-90W, while the flatpak one is 70W, and with OpenGL disabled it's about 60W. So it's better, but not as good as OpenGL disabled.

I'm looking at my power meter when doing these tests. The UI is fluent in all tests for my machine with a R7 5700X CPU.

**Comment by ilya-fedin** [comment]
@lilydjwg so you don't have this issue? This issue is about laggy UI, please look the video in the forth comment

**Comment by lilydjwg** [comment]
I don't have it on this machine. If my cpu weren't powerful enough, I'd have a laggy ui.

My original issue was closed as a dup: https://github.com/telegramdesktop/tdesktop/issues/27147#issuecomment-1826768802.

**Comment by ilya-fedin** [comment]
Maybe the flatpak one is good enough? I don't think it's possible to make any better, the rest of usage is likely to the simple fact it's CPU rendering.

**Comment by lilydjwg** [comment]
> Maybe the flatpak one is good enough?

I'm not going to do a laptop battery life test any time soon, so wait for others to comment?

Is there any downsides to disable the OpenGL option?

**Comment by ilya-fedin** [comment]
> Is there any downsides to disable the OpenGL option?

worse performance of the media viewer and video calls/chats

**Comment by ilya-fedin** [comment]
@K900 @LDVSOFT @PoSayDone @andresilva @pranayadmn @vitordelucca do you have the issue with any of the official builds (static binary/flatpak/snap) of 4.14.3? If yes, does the above flatpak test build have the issue?

@gamelaster do you have the issue on the flarpak test build?

**Comment by gamelaster** [comment]
@ilya-fedin I am using 4.14.3 from Arch Linux repositories. I tried the flatpak test build, when disconnecting the internet to turn on the disconnect animation, the CPU consumption is lower:
![image](https://github.com/telegramdesktop/tdesktop/assets/1665373/6774f120-d628-4b58-9785-71886abbdd5f)
(before it was 20% - 40%)

Although, still the animations are sluggish, and the CPU usage sometimes peaks to 70%. Also, I am using KDE Plasma. If any further assist is needed, let me know. Thanks!

**Comment by ilya-fedin** [comment]
> I am using 4.14.3 from Arch Linux repositories

Unofficial builds like the ones from distro repositories won't get any improvements cause the bottlenecks seem to be inside Qt and I solve them by patching Qt...

**Comment by gamelaster** [comment]
@ilya-fedin makes sense. Although, sadly, using flatpack version with (I hope so) patched Qt did not gave any improvements :(

**Comment by ilya-fedin** [comment]
@gamelaster I wonder why it's not better given that CPU consumption is way lower. Maybe you can write a screen record how is it for you?

**Comment by ilya-fedin** [comment]
I've compared Qt's WAYLAND_DEBUG log and the Firefox's one, it indeed looks like Qt is unable to tell the damage to the compositor with OpenGL for whatever reason while Firefox does this just fine somehow. I'm quite confused how that happens given that damage seem to be handled internally by mesa during the eglSwapBuffers call...

**Comment by ilya-fedin** [comment]
Looks like Qt doesn't even collect this information in OpenGL mode so maybe it's cheap to redraw entire window when the buffer is on the GPU...

**Comment by ilya-fedin** [comment]
@gamelaster what gpu do you have btw?

**Comment by lilydjwg** [comment]
But the high usage is on CPU, not GPU, so compositing should not be the issue (compositing has to be done somewhere anyway). Does Telegram redraw all its widgets when Qt decides to redraw the entire window?

**Comment by lilydjwg** [comment]
I did flamegraphs with OpenGL enabled and disabled. The CPU is used inside Telegram itself.

With OpenGL enabled:
![opengl-yes](https://github.com/telegramdesktop/tdesktop/assets/440661/25d849ef-0051-4cb3-a03c-7b0c9e05d9d0)

With OpenGL disabled:
![opengl-no](https://github.com/telegramdesktop/tdesktop/assets/440661/369726ac-87ed-4184-a404-8cad6358f185)

**Comment by ilya-fedin** [comment]
> But the high usage is on CPU, not GPU

As far as I understand it's not? The person says it's 3% down from 20%-40% but it's still sluggish what makes me think CPU usage is not relevant to the problem.

> Update: or perhaps the buffer is still passed via memory when OpenGL enabled so the CPU usage is due to large buffer copies?

That would mean something is really wrong in the OpenGL driver or the llvmpipe driver is used. For me it uses dmabuf with the amdgpu driver (I see via `WAYLAND_DEBUG=1`).

**Comment by ilya-fedin** [comment]
> This is done with the Linux distro package because I have debug symbols for it.

It doesn't seem to be readable...

**Comment by ilya-fedin** [comment]
> Does Telegram redraw all its widgets when Qt decides to redraw the entire window?

If Qt decides to redraw entire window, then yes, it will send a paint event of entire window to Telegram. But that usually happens only when resizing, the rest of the time widgets outselves trigger the redraw.

**Comment by lilydjwg** [comment]
> As far as I understand it's not? The person says it's 3% down from 20%-40% but it's still sluggish what makes me think CPU usage is not relevant to the problem.

I'm sorry you're right. The high CPU usage I saw is what you've just fixed via the Qt patch. With the flatpak one, I don't see high CPU usage, but only a bit of extra GPU usage (nvtop shows GPU power usage goes from ~4W to ~8W).

**Comment by ilya-fedin** [comment]
I wouldn't be surprised if the problem now remain only with Intel GPU which are known to have bad performance on Wayland due to the lack of tripple buffering in the compositors so I really would like to hear feedback from everyone affected in this thread, especially the ones that have NVIDIA/AMD rather then Intel GPU.

**Comment by K900** [comment]
4.14.3 seems fine, or at least much improved, on NixOS staging-next. I'm not entirely sure why...

**Comment by ilya-fedin** [comment]
Looks like there will be 4.14.4 so all official builds (static binary, flatpak, snap) will have the patch today

**Comment by gamelaster** [comment]
@ilya-fedin I have AMD Radeon Pro W5500 GPU, and I am running Arch Linux with Linux 6.6.9, with KDE Plasma on Wayland. Also, I have 2560x1440 monitor, thus I have enabled upscale of UI. I recorded video with my phone, so I am sure that recording of the video doesn't impact performance. The app on the video is the one from flatpak you provided. On end of the video, I am disabling network, so reconnect animation is visible, which causes high cpu usage as well.
(also, not sure if it is visible, but menu open animation is stuttering)
https://gami.ee/files/telegram.mp4

**Comment by ilya-fedin** [comment]
@gamelaster I've looked your video and both animations and the CPU usage looks OK to me

**Comment by gamelaster** [comment]
@ilya-fedin hmm, this is strange. I used Telegram Desktop for several months with all animations enabled, and did not noticed an issue, only after some update, it started to be more stuttering. So, I guess there is chance that something is messed up at my side. Although, for sure, the flatpack version performs better and consumes less CPU, than the one from Arch repo. Thank you for your time.

**Comment by ilya-fedin** [comment]
@LDVSOFT @PoSayDone @andresilva @pranayadmn @vitordelucca I'm still waiting for your feedback

**Comment by andresilva** [comment]
Sorry but I don't have access to my desktop right now (which is where I'm sure the problem is happening). I don't see any problem on my laptop (with intel gpu) on 4.14.3. I'll give you an update on the weekend after I test on my desktop.

**Comment by pranayadmn** [comment]
> @LDVSOFT @PoSayDone @andresilva @pranayadmn @vitordelucca I'm still waiting for your feedback

Unfortunately, I'm back to college since my spring sem started. I only have my Windows laptop with me right now :P

**Comment by PoSayDone** [comment]
> ... I'm still waiting for your feedback

No stutters, everything works fine :)

**Comment by LDVSOFT** [comment]
4.14.4 is way better. At some point it still felt as a short lag but it's not "UI is 10 fps & delayed for half .2 seconds behind the mouse", so definitely a major improvement. Quite interested what was changed.

**Comment by ilya-fedin** [comment]
> Quite interested what was changed.

if 4.14.3 was bad for you then it's Qt patches to get rid of unneeded buffer copies

**Comment by andresilva** [comment]
Unfortunately still happening for me on 4.14.4 😞 Disabling OpenGL fixes it though.

**Comment by ilya-fedin** [comment]
@andresilva I hope you're using not a distro build? They won't have improvements.

**Comment by andresilva** [comment]
No I'm not, sorry I missed that, using the flatpak is not an option for me. Either way disabling OpenGL fixes it which is good enough for me (I think it didn't fix it last time I tried, or maybe I mixed it up).

**Comment by ilya-fedin** [comment]
@andresilva flatpak is not the only choice, there are also static binary and snap. Snap doesn't seem to have 4.14.4 for whatever reason, though.

**Comment by andresilva** [comment]
I'm using NixOS so I really want to install everything with the package manager, either way I tried the static binary (I just needed to manually patch the dynamically linked libraries) and I can report that with that binary the issue is solved even with OpenGL enabled 👍.

**Comment by ilya-fedin** [comment]
Sorry folks, the fix is reverted and the next version will have bad performance again

**Comment by ljuzig** [comment]
Media Viewer is completely broken on Gnome Wayland too. When you click over an image, it does not shows anything, and do not let you close the "invisible window", preventing Telegram from quitting. You have to manually kill the process. Using QT_QPA_PLATFORM=xcb fixes this problem but introduces other issues with fractional scaling (blurred font), stuttering and other stuff.

**Comment by ljuzig** [comment]
@ilya-fedin are you sure it's completely unrelated? It was ok a while ago (and it is with no wayland), I think before around 4.14 so I thought it might be related or useful to this issue. I haven't yet checked if it's ok after the reverse.

**Comment by ilya-fedin** [comment]
@ljuzig the fix is reverted in 5.0, not 4.14. I don't see how that could be related.

**Comment by ljuzig** [comment]
That's the point. Version <= 4.14 was ok (don't really know the exact version), then something broke it, supposedly something regarding this fix? I don't see any commit mentioned here so I cannot check myself what actually changed and when. Maybe 5.0 revert will make this work again? Cannot test right now. I thought it was worth mentioning since apparently applying the flags suggested in this issue fixes the problem, but it may very well be something unrelated.

**Comment by ilya-fedin** [comment]
> Version <= 4.14 was ok (don't really know the exact version), then something broke it, supposedly something regarding this fix?

I don't see how this could be related if you don't even know exact version. 4.14 was released in January, do you have the problem since January? Viewer not appearing sounds very different to this probklem which is about entire application UI being lagged, I don't see any connection.

> I don't see any commit mentioned here so I cannot check myself what actually changed and when.

This is the reverting commit: https://github.com/desktop-app/patches/commit/ce5098d9bd638ec264fd8cf69ef3ef31d8139d18

> Maybe 5.0 revert will make this work again? 

No, I don't think your problem is related to this issue.

**Comment by ljuzig** [comment]
Ok, ty. Gonna open a new issue then.

**Comment by LDVSOFT** [comment]
Is there any reason behind the revert?

**Comment by ilya-fedin** [comment]
@LDVSOFT #27736

**Comment by ilya-fedin** [comment]
Hmm, maybe replacing the underlying backing store with the one that uses owning QImage would help (e.g. QRhiBackingStore)...
